### PR TITLE
Pad sequences and Pack sequences

### LINKED ISSUE #1128: Pad a list of tensors
Taking in a list of tensors, can we have a function to pad them? This is a useful feature to process sequences of different lengths and process mini batches. 

I think the documentation of pad_packed_sequence and torch.nn.utils.rnn.pack_padded_sequence are a little confusing http://pytorch.org/docs/nn.html?highlight=pad_packed#torch.nn.utils.rnn.pad_packed_sequence maybe adding some examples could help the explanation.

### PR DESCRIPTION
**Pull request by hhsecond**
PR resolves  [#1128](https://github.com/pytorch/pytorch/issues/1128)

- Handling sort internally was supposed to be in this PR but I couldn't decide how we keep track of the actual order in case user needs the sequence in that order.

### CODE REVIEW

File: torch/nn/utils/rnn.py [hunk 1]
```diff
@@ -123,3 +124,106 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+    stack all the sequences on zeroth dimension. For example, if the input is
+    list of sequences with size `` Lx*`` and if batch_first is False, the
+    output will be of size `` TxBx* `` and if batch_first is True,
+    output will be of size ``BxTx* ``. The ``pad_sequence`` accepts list of
+    sequences and its lengths which should be sorted in decreasing order
+
+    B is batch size
+    T is length longest sequence
+    L is length of the sequence
+    * is any trailing dimension including zero
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: it should be "any number of trailing dimensions, including none"

File: torch/nn/utils/rnn.py [hunk 2]
```diff
@@ -123,3 +124,106 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+
+    >>> from torch.nn.utils.rnn import pad_sequence
+    >>> a = Variable(torch.ones(25, 300))
+    >>> b = Variable(torch.ones(15, 300))
+    >>> c = Variable(torch.ones(22, 300))
+    >>> pad_sequence([a, b, c], [25, 15, 22]).size()
+    torch.Size([25, 3, 300])
+
+    Note:
+        This function returns a Variable of size LxBx* or BxLx* where L is the
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: You used `T` to denote the max of lengths previously.

File: torch/nn/utils/rnn.py [hunk 3]
```diff
@@ -123,3 +124,106 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+        length of longest sequence (lengths[0])
+
+    Arguments:
+        sequences (list(Variable)): list of variable length sequences.
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: nit: `list[Variable]`

File: torch/nn/utils/rnn.py [hunk 4]
```diff
@@ -123,3 +124,106 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+        lengths (list[int]): list of sequences lengths of each batch element.
+        batch_first (bool, optional): if True, the input is expected in Bx*x*
+            format.
+
+    Returns:
+        Variable of size ``seq_len x len(sequences) x * ``
+            if batch_first = False
+        Variable of size ``len(sequences) x seq_len x * `` otherwise
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: I think it's best to stick to `T` instead of `seq_len`

File: torch/nn/utils/rnn.py [hunk 5]
```diff
@@ -123,3 +124,106 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+            prev_l = length
+            padding_shape = [lengths[0] - length] + list(variable.size()[1:])
+            filler = Variable(torch.Tensor(*padding_shape).type_as(variable.data).zero_())
+            out_variable.append(torch.cat((variable, filler)))
+        else:
+            out_variable.append(variable)
+    if batch_first:
+        return torch.stack(out_variable)
+    else:
+        return torch.stack(out_variable).transpose(0, 1)
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: this if can be replaced with `return torch.stack(out_variable, 0 if batch_first else 1)`

File: torch/nn/utils/rnn.py [hunk 6]
```diff
@@ -123,3 +124,106 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+    Returns:
+        Variable of size ``seq_len x len(sequences) x * ``
+            if batch_first = False
+        Variable of size ``len(sequences) x seq_len x * `` otherwise
+    """
+
+    if len(lengths) != len(sequences):
+        raise ValueError("number of elements in lengths and sequences didn't match")
+
+    out_variable = []
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: I think it would be more efficient to pre-allocate and zero-fill that Variable, and then copy the contents of sequences into it (with appropriate slicing). This will also make it easier to make this function GPU-compatible (right now it's not, because `filler` might end up on a wrong device).

File: test/test_nn.py [hunk 7]
```diff
@@ -2164,6 +2164,83 @@ def test_invalid_dropout_p(self):
...
+
+        # batch_first = false
+        padded = rnn_utils.pad_sequence([a, b, c], [3, 2, 1])
+        self.assertEqual(padded, expected.transpose(0, 1))
+
+        # more dimensional
+        sequences = []
+        lengths = []
+        maxlen = 9
+        for i in reversed(range(maxlen + 1)[1:]):
```
Comment (apaszke) [inline test/test_nn.py]: This iterator can be rewritten as `range(maxlen, 0, -1)`

File: test/test_nn.py [hunk 8]
```diff
@@ -2164,6 +2164,83 @@ def test_invalid_dropout_p(self):
...
+            seq_len = i * i
+            lengths.append(seq_len)
+            sequences.append(Variable(torch.rand(seq_len, 400)))
```
Comment (apaszke) [inline test/test_nn.py]: can you make it smaller than 400? 5 should be enough

File: test/test_nn.py [hunk 9]
```diff
@@ -2164,6 +2164,83 @@ def test_invalid_dropout_p(self):
...
+        packed = rnn_utils.pack_sequence(sequences, lengths)
+        unpacked = rnn_utils.pad_packed_sequence(packed, batch_first)
+        self.assertEqual(padded, unpacked[0])
+
+        batch_first = False
+        padded = rnn_utils.pad_sequence(sequences, lengths, batch_first)
+        packed = rnn_utils.pack_sequence(sequences, lengths)
+        unpacked = rnn_utils.pad_packed_sequence(packed, batch_first)
+        self.assertEqual(padded, unpacked[0])
+        #####################################
```
Comment (apaszke) [inline test/test_nn.py]: can you please remove those separators?

File: test/test_nn.py [hunk 10]
```diff
@@ -2164,6 +2164,83 @@ def test_invalid_dropout_p(self):
...
+        for i in reversed(range(maxlen + 1)[1:]):
+            seq_len = i * i
+            lengths.append(seq_len)
+            sequences.append(Variable(torch.rand(seq_len, 400)))
+
+        # compatibility with other utilities
+        batch_first = True
+        padded = rnn_utils.pad_sequence(sequences, lengths, batch_first)
+        packed = rnn_utils.pack_sequence(sequences, lengths)
+        unpacked = rnn_utils.pad_packed_sequence(packed, batch_first)
```
Comment (apaszke) [inline test/test_nn.py]: I think it'd be bette to do `pack_sequence` and compare that to `pad_sequence` + `pack_sequence` (instead of `pad_sequence` vs `pack` + `unpack`). Or even better - do both!

File: test/test_nn.py [hunk 11]
```diff
@@ -2164,6 +2164,83 @@ def test_invalid_dropout_p(self):
...
+        a = Variable(torch.Tensor([1, 2, 3]))
+        b = Variable(torch.Tensor([4, 5]))
+        c = Variable(torch.Tensor([6]))
+        packed = rnn_utils.pack_sequence([a, b, c], [3, 2, 1])
+        expected = torch.Tensor([1, 4, 6, 2, 5, 3])
+        self.assertEqual(packed.batch_sizes, [3, 2, 1])
+        self.assertEqual(packed.data.data, expected)
+
+        # more dimensions
+        ####################################
```
Comment (apaszke) [inline test/test_nn.py]: Can you make this test more than one dimension? Just do a for loop for 1, 2, 3 trailing dims.

**Review by apaszke — COMMENTED** [review]
That looks really good. I left some minor comments. Thanks for the PR!

File: torch/nn/utils/rnn.py (hunk 6 above)
Comment (hhsecond) [reply torch/nn/utils/rnn.py]: Wouldn't the ``type_as`` in ``filler`` creation make it GPU compatible?

Comment (apaszke) [reply torch/nn/utils/rnn.py]: It kind of will, but not completely - the tensor will be on a GPU, but on the "current GPU" set in the context, not the one of the inputs. You have to use `.new` if you want to retain the device.

**Comment by apaszke** [comment]
Definitely `tensor.new(shape).zero_()`. THe other one won't preserve the device (as I explained before).

**Comment by soumith** [comment]
@apaszke is it good to merge?

**Comment by apaszke** [comment]
I have this PR open in a tab all the time... Will try to review soon, sorry.

File: torch/nn/utils/rnn.py [hunk 12]
```diff
@@ -123,3 +124,126 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+            prev_l = length
+            if length < max_len:
+                prev_l = length
+                padding_dims = [lengths[0] - length] + trailing_dims
+                filler = Variable(variable.data.new(*padding_dims).zero_())
+                out_variable[i] = torch.cat((variable, filler))
+            else:
+                out_variable[i] = variable
+    else:
+        # repeating the same logic but with T as first dimension
```
Comment (fmassa) [inline torch/nn/utils/rnn.py]: I believe you can remove this code duplication. For the indexing, you could use `.select` and specify the dimension to index in the condition. What do you think?

Comment (hhsecond) [reply torch/nn/utils/rnn.py]: I was looking for something similar but couldn't find anything. I guess ``.select`` makes a copy and I won't be able to assign the value to that slice. Am I not correct?
for ex. below case my ``a`` won't change at all
``
>>> a = Variable(torch.rand(2,3))
>>> a.select(1, 1).data = torch.zeros(2)
``

Comment (fmassa) [reply torch/nn/utils/rnn.py]: `select` does not perform a copy, but returns an object that shares the same storage. But maybe using `select` might become cumbersome.
If it is easier, you could instead index with tuples (which was used internally in an old version of `select`).
Something like
```python
a = Variable(torch.rand(2, 3))
index = (i, ) if batch_first else (slice(None), i)
a[index] = ...
```

File: torch/nn/utils/rnn.py [hunk 13]
```diff
@@ -123,3 +124,126 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
...
+        for i, variable, length in zip(range(len(lengths)), sequences, lengths):
+            # temperory sort check, can be removed when we handle sorting internally
+            if prev_l < length:
+                    raise ValueError("lengths array has to be sorted in decreasing order")
+            prev_l = length
+            if length < max_len:
+                prev_l = length
+                padding_dims = [lengths[0] - length] + trailing_dims
+                filler = Variable(variable.data.new(*padding_dims).zero_())
+                out_variable[i] = torch.cat((variable, filler))
```
Comment (fmassa) [inline torch/nn/utils/rnn.py]: Question: instead of performing the concatenation, can't we instead slice `out_variable` and only copy `variable`?
Something like (untested)
```python
out_variable[i, :length] = variable
```
?

Comment (apaszke) [reply torch/nn/utils/rnn.py]: Yes, it should definitely be changed to slicing. The whole point of pre-allocating a variable was to avoid these `cat`s

File: torch/nn/utils/rnn.py (hunk 12 above)
Comment (apaszke) [reply torch/nn/utils/rnn.py]: Yup, there should be a single code path for both cases. There's no reason to separate them. I think it's best to use `select`.

File: torch/nn/utils/rnn.py (hunk 13 above)
Comment (hhsecond) [reply torch/nn/utils/rnn.py]: that makes a lot of sense, how didn't I see that!!
Will do an updated PR soon.
B/w @apaszke, @fmassa , the assigning value to selected ``variable.data`` doesn't change the actual parent object, why is that happening since they shared the same storage?
`
>>> a = Variable(torch.rand(2,3))
>>> a
Variable containing:
 0.6392  0.6609  0.3849
 0.2233  0.3120  0.9595
[torch.FloatTensor of size 2x3]

>>> a.select(1, 1).data = torch.zeros(2)
>>> a
Variable containing:
 0.6392  0.6609  0.3849
 0.2233  0.3120  0.9595
[torch.FloatTensor of size 2x3]

`

Comment (apaszke) [reply torch/nn/utils/rnn.py]: `.data` modifications happen outside of autograd, and are invisible to it. That's why you should never do these things if you want them differentiated. Don't unpack the data here, modifying Variables in-place is ok.

**Comment by hhsecond** [comment]
@fmassa @apaszke Updated

File: test/test_nn.py [hunk 14]
```diff
@@ -2164,6 +2164,84 @@ def test_invalid_dropout_p(self):
...
+        # unsorted sequences should raise exception
+        self.assertRaises(
+            ValueError, lambda: rnn_utils.pad_sequence([b, a, c], [2, 3, 1]))
+
+    def test_pack_sequence(self):
+        def _compatibility_test(sequences, lengths, batch_first):
+            padded = rnn_utils.pad_sequence(sequences, lengths, batch_first)
+            packed = rnn_utils.pack_sequence(sequences, lengths)
+            unpacked = rnn_utils.pad_packed_sequence(packed, batch_first)
+            self.assertEqual(padded, unpacked[0])
```
Comment (apaszke) [inline test/test_nn.py]: Why is this `[0]` necessary? This seems incorrect to me

**Review by apaszke — COMMENTED** [review]
LGTM, but it seems that there's an error in the test. Can you please explain to me why is this select necessary?

Sorry for having you wait so long. 😕

File: test/test_nn.py (hunk 14 above)
Comment (elanmart) [reply test/test_nn.py]: @apaszke isn't `unpacked` a tuple of `(padded_var, seq_lengths)`?

Comment (apaszke) [reply test/test_nn.py]: Ah, that makes sense! Thanks Marcin!

**Review by soumith — APPROVED** [review]

**Comment by apaszke** [comment]
Don't merge yet. I'm fixing some absolutely minor things in the docstrings and simplifying the code a tiny bit.

**Comment by elanmart** [comment]
I have a question regarding this PR:

couldn't the `pack_sequence` function take `lengths=None` as a default parameter, and just compute those internally as `lengths=[v.size(0) for v in sequences]`?

**Comment by apaszke** [comment]
That's true. I'll simplify the code to reflect that.

**Comment by elanmart** [comment]
Also, the docstring doesn't mention that this `List[Variables]` should be sorted, and I believe it should?

**Comment by apaszke** [comment]
It does mention that they should be sorted in the order of decreasing length

**Comment by elanmart** [comment]
I think `pad_sequence` does, and `pack_sequence` doesn't?

**Comment by apaszke** [comment]
Alright, I've simplified a few things and mentioned the requirements in the comments. Let me know if you see anything else.

One thing I just started wondering about is whether stacking is what we actually want. An alternative I see would be to require inputs to be `L x 1 x *` if `batch_first` is false, and `1 x L x *` otherwise. Not sure what plays better with other parts of the API.

**Comment by hhsecond** [comment]
@apaszke Thanks for the checks and modifications. Why do you think stacking would be better than pushing to the placeholder. The current approach had given us potential performance gain. I'll be happy to change it to stacking if that's what we want

**Comment by apaszke** [comment]
@hhsecond sorry that's not what I meant. We should definitely continue with the approach we took now, my only concern is what kinds of input sizes should we expect from the users. It's purely a matter of choosing the right interface - there are very little changes required in the implementation. When I said "stacking" I meant that we add a new dimension, instead of "concatenating" (even though we don't use `torch.cat` directly) along another one.

**Comment by jekbradbury** [comment]
Adam brings up a good point; I don't have a strong immediate feeling one way or another so let's try to look at different use cases:
First, the most common use of these functions will probably be in preparing data. There, code often (but not always) operates on sequences of token IDs (which have no feature dimension) rather than sequences of feature vectors. When that's the case existing code sometimes performs padding directly from a sequence of Python lists. So it would definitely be helpful to make sure these functions work on data with no feature dimension, and it could potentially also be worthwhile to make them work on inputs whose time dimension is encoded as a list.
Other than that, people writing dataloaders are likely to marginally prefer providing a sequence of `L x F` tensors to having to unsqueeze each of them (the tensors might be coming from numpy or a scikit-learn dataloader, where PyTorch's convention of using a batch-size-one tensor to denote a single example is somewhat foreign).
The other use that I can think of is people either writing complicated manually-batched networks where flow switches back and forth between sections batched in different ways or people writing libraries that do the same thing automatically. While most of the time these people will actually want to use `pad_packed_sequence` and `pack_padded_sequence`, they'd want to use these functions if they have completely unbatched sections of code. And within those sections they'd want to use PyTorch-style unbatched tensors, i.e. tensors with batch size 1.
So there are reasons for each choice but the reasons for stacking are probably less esoteric and more frequently encountered.

**Comment by apaszke** [comment]
That's super helpful James, thanks a lot. Let's go with stacking

**Comment by hhsecond** [comment]
@apaszke Are we merging this now? Making it compatible with python list should not be a big deal but then I am not sure how to decide what would be the tensor type in case of token IDs and looks like we need to handle 3 different input kinds then
- List[Tensors/Variables]
- List[List[Tensors/Variables]]
- List[List[Int/Float]]

**Comment by apaszke** [comment]
Let's limit it to Variables only for now

**Comment by apaszke** [comment]
Thank you @hhsecond!

**Comment by williamFalcon** [comment]
@apaszke this needs testing. pad_sequence points to duplicate instance of the same Variable and crashes.

**Comment by apaszke** [comment]
@williamFalcon it is tested (see the diff)

**Comment by williamFalcon** [comment]
@apaszke ok. Just ran a basic test and verified it's indeed crashing. I already created an issue and submitted a PR. #4381

**Comment by nikhilweee** [comment]
Is this included in 0.3.1? For now `from torch.nn.utils.rnn import pad_sequence` throws an error. Is this the intended behaviour?
